test-basic/documents-data-movement-readAll.js

```diff
@@ -216,19 +235,179 @@ describe('data movement readAll', function() {
+                        documents.length.should.equal(1000);
+                        return 10;
+                    })
+                })).on('error', function(err){
+                    err.toString().should.equal('Error: onBatchError should return null, empty array or a replacement array.');
+                    done();
+                });
+            })
+            .catch(err=> done(err));
+    });
+
+    it('should readAll documents with onBatchError option returning null', function(done){
+        restAdminDB.config.transforms.write(xqyTransformName, 'xquery', fs.createReadStream(xqyTransformPath))
+            .result(function(response){
+                uriStream.pipe(dbWriter.documents.readAll({
+                    batchSize:1000,
+                    transform: ['tested1'],
+                    onBatchError: ((progressSoFar, documents, error) => {
+                        progressSoFar.docsReadSuccessfully.should.be.equal(0);
+                        progressSoFar.docsFailedToBeRead.should.be.equal(1000);
+                        error.body.errorResponse.messageCode.should.be.equal('RESTAPI-INVALIDREQ');
+                        documents.length.should.equal(1000);
+                        return null;
+                    }),
+                    onCompletion: ((summary) => {
+                        summary.docsReadSuccessfully.should.be.equal(0);
+                        summary.docsFailedToBeRead.should.be.equal(10000);
+                        summary.timeElapsed.should.be.greaterThanOrEqual(0);
+                        done();
+                    })
+                }));
+            })
+            .catch(err=> done(err));
+    });
+
+    it('should readAll documents with onBatchError option returning replacement batch', function(done){
+        let onBatchErrorStream = new Stream.PassThrough({objectMode: true});
+        onBatchErrorStream.push('');
+        onBatchErrorStream.push(null);
+        onBatchErrorStream.pipe(dbWriter.documents.readAll({
+            batchSize:10,
+            onBatchError: ((progressSoFar, documents, error) => {
+                let replacementBatch = [];
+                for (let i = 0; i < 10; i++) {
+                    replacementBatch.push('/test/dataMovement/requests/readAll/' + i + '.json');
+                }
+
+                onBatchErrorStream = new Stream.PassThrough({objectMode: true});
+                onBatchErrorStream.push(null);
+                return replacementBatch;
+            }),
+            onCompletion: ((summary) => {
+                summary.docsReadSuccessfully.should.be.equal(10);
+                summary.docsFailedToBeRead.should.be.equal(0);
+                summary.timeElapsed.should.be.greaterThanOrEqual(0);
+                done();
+            })
+        }));
+    });
+
+    it('should readAll documents with consistentSnapshot option as true', function(done){
```

Comment:
A good test for a consistent snapshot would be to update a document after the first document is read but before the modified document is read.

The output should have the old content for the modified document because the timestamp of the modification is after the consistent timestamp.

Reply:
I will work on this.